### ISSUE TITLE
Modifying sources in catalogs and updating old drivers to v2

### ISSUE THREAD

**Issue by kthyng** [intake/intake#826]
Sorry to open so many new issues! I have another question.

I am working with [intake-erddap](https://github.com/axiom-data-science/intake-erddap) for the first time with intake v2. It still works fine, except in my previous workflow that I need to use again, I modified the sources from the catalog after they are created. This previously was tricky (I couldn't just change the metadata, I had to change the metadata from .describe(), but then it worked), but I tried this workflow again and I wasn't able to get the modifications to stick this time.  According to your note [here](https://intake.readthedocs.io/en/latest/tour2.html#migration-from-v1), I maybe need to change to a different base class of `intake.readers.entry:Catalog` which I see also then has methods of `add_entry()` and `delete()`. Am I correct in guessing these might fix my issue of wanting to be able to modify sources in that I could modify them and then delete the original source and add the new version into the catalog? I have spent the afternoon trying to update to the new base class and haven't been successful so far so wanted to make sure I am on the right track before spending more time on it.

**Comment by martindurant** [comment]
I am not quite following what you want to do. We are talking about adapting https://github.com/axiom-data-science/intake-axds/blob/main/intake_axds/axds_cat.py#L28 to V2? You would probably end up doing something like [this](https://github.com/intake/intake/blob/master/intake/readers/catalogs.py#L89) where you create the readers (or reader-descriptions) directly. The following case in the same file for STAC has a more involved workflow, assigning metadata and choosing different readers for each entry. 

So, you either make readers and assign them into the catalog object you are making (`cat[name] = reader`) or you make description objects directly, or you can even make "lazy" entries like the Tiled reader, where you only look up entries on demand. Many options! The metadata in every case is just a dictionary, so you can always edit it in-place.

**Comment by kthyng** [comment]
No, I am working on https://github.com/axiom-data-science/intake-erddap/blob/main/intake_erddap/erddap_cat.py#L37. I had eventually found my way to that example actually. I think I have something set up in the vain of `SQLAlchemyCatalog` (https://github.com/intake/intake/blob/master/intake/readers/catalogs.py#L71) but I have been stuck with the `DataDescription` among other things. I think it should be working but it isn't. Now I think it is from not having my package installed correctly. I switched my `erddap.py` sources to inheriting from `from intake.readers.datatypes import BaseData` with `erddap_cat.py` inheriting from `from intake.readers.readers import BaseReader`, and I switched the entry_points as you stated in the developer docs note, along with trying to follow your note with `intake.conf["extra_imports"]` (for which I have in my `__init__` the following:
```
import intake
intake.conf["extra_imports"] += ["intake_erddap"]
```
)
but once I uninstalled and reinstalled locally with pip after switching my entry points to  "intake.imports" instead of  "intake.drivers", intake couldn't find my catalog object anymore for a call like this:
```
catalog = intake.open_erddap_cat()
```
The catalog call worked once I switched the catalog entry point back to "intake.drivers". However, I don't seem to be able to get my two sources to be found by intake either way — I think that is at least one problem I am having. Am I missing something?

Here is a more complete snippet of my setup.py:
```
setup(
...
    packages=["intake_erddap"],
    package_data={"": ["*.csv", "*.yml", "*.html"]},
    entry_points={
        "intake.imports": [
            "tabledap = intake_erddap.erddap:TableDAPSource",
            "griddap = intake_erddap.erddap:GridDAPSource",
            # "erddap_cat = intake_erddap.erddap_cat:ERDDAPCatalogReader",
        ],
        "intake.drivers": [
            "erddap_cat = intake_erddap.erddap_cat:ERDDAPCatalogReader",
        ]        
    },
    include_package_data=True,
    install_requires=requires,
...
)
```

**Comment by martindurant** [comment]
```
import intake
intake.conf["extra_imports"] += ["intake_erddap"]
```

This will have no effect, because the importing happens during `import intake`, and you can only set the config afterward. If you want to use the value, you'll need
```
import intake
intake.conf["extra_imports"] += ["intake_erddap"]
intake.conf.save()
```
(and then it will be picked up the next time around).


```
        "intake.imports": [
            "tabledap = intake_erddap.erddap:TableDAPSource",
            "griddap = intake_erddap.erddap:GridDAPSource",
            # "erddap_cat = intake_erddap.erddap_cat:ERDDAPCatalogReader",
        ],
```
if the last line (uncommented) doesn't cause an import of `intake_erddap.erddap_cat` along with `import intake`, then we have a bug. You could uncomment it, and put a print() statement in the global scope of the module to test.


Note that there is another config value, "allow_import", which will turn off the whole mechanism if set to False, and "import_block_list", a list of things *not* to import - but I don't suppose you have changed these. I would get around to writing long-form descriptions of what all the config values do.

**Comment by kthyng** [comment]
Ok, I put this:
```
import intake
intake.conf["extra_imports"] += ["intake_erddap"]
intake.conf.save()
```
into the `__init__` of `intake-erddap` and uninstalled and reinstalled (`pip install .`) and imported intake a few times in different kernels. It was not clear to me that is what you meant by "and saving", but I did save my file! 

> if the last line (uncommented) doesn't cause an import of intake_erddap.erddap_cat along with import intake, then we have a bug

No I have to use the "intake.drivers" version to have `erddap_cat` found by intake. However, I haven't been able to get intake to find "intake_erddap.erddap:TableDAPSource" although I thought it would work using these entry points — is that not the case? 

> put a print() statement in the global scope of the module to test

How do I do this?

I simply changed the ERDDAPSource inherited class from `base.DataSource` to `BaseData` — is that a problem to not update more? Similar to for the catalog class to using `BaseReader`, though I have changed more in that class. Maybe it would be better to go back to what I had before from v1? It was actually working mostly fine!

**Comment by martindurant** [comment]
I created https://github.com/intake/an_intake_project as a minimal example of how this is supposed to works and ... it doesn't work! The importing does indeed happen, but the classes don't get imported:
```
>>> import intake
import intake project
Importing intake_project:MyClass as part of processing intake entrypoints failed
(partially initialized module 'intake' has no attribute 'BaseData' (most likely due to a circular import))
```

I'll fix it!

**Comment by martindurant** [comment]
https://github.com/intake/intake/pull/829

**Comment by martindurant** [comment]
Following that PR, the following works:
```
>>> import intake
import intake project
>>> any("ExampleData" in str(_) for _ in intake.readers.utils.subclasses(intake.BaseData))
True
```

but if I do
```
>>> intake.conf["import_block_list"] = ["intake_project"]
>>> intake.conf.save()
```
then next time there is no print and ExampleData is not registered.

**Comment by kthyng** [comment]
```
>>> intake.conf["import_block_list"] = ["an_intake_project"]
>>> intake.conf.save()
```

It sounds like you are putting this directly in your local `intake` and I don't see it in `an_intake_project` but how can that work to distribute it to other users? Do you just run those commands once and then you're done or do you stash that code somewhere forever? And wouldn't you need a `+=` if you add more than one intake driver over time? So I guess my question is: where do you put these lines exactly?

> then next time there is no print and ExampleData is not registered.

We want `ExampleData` to be registered right? I am trying to make sure I am following you correctly here. So are you saying the next time you imported intake it did or didn't work to import your driver?

**Comment by kthyng** [comment]
Ok I just realized you were using a different config option "import_block_list" (now that makes more sense) — I thought we were still talking about "extra_imports" which I may not be using correctly. Is "extra_imports" necessary to use for the entry_points to be found by intake, and if so where exactly do I need to put those config lines?

**Comment by martindurant** [comment]
> Is "extra_imports" necessary to use for the entry_points to be found by intake

No, it's a potential alternative for packages that lack entrypoints

> where exactly do I need to put those config lines

Again, you would edit the `intake.conf` object and save it, to have it take effect next time. Maybe also turn on the logger "intake" which maybe gives you more information on what's going wrong.

**Comment by kthyng** [comment]
> No, it's a potential alternative for packages that lack entrypoints

Ah, ok.

Now I am able to get both of the following to return True: 

```
any("TableDAPSource" in str(_) for _ in intake.readers.utils.subclasses(intake.BaseData))
any("erddap_cat" in str(_) for _ in intake.readers.utils.subclasses(intake.BaseReader))
```

However, this syntax still isn't working `intake.open_erddap_cat()` where the `open_*` catalog automatically works once the catalog name is findable. Is that meant to work?

Error:
```
AttributeError: Unknown open method 'open_erddap_cat'. Do you need to install a new driver from the plugin directory? https://intake.readthedocs.io/en/latest/plugin-directory.html
Registered opener methods: ['open_coops-dataframe', 'open_coops-xarray', 'open_coops_cat', 'open_csv', 'open_jsonfiles', 'open_numpy', 'open_textfiles', 'open_tiled_cat', 'open_zarr', 'open_netcdf', 'open_opendap', 'open_rasterio', 'open_remote-xarray', 'open_xarray_image', 'open_yaml_file_cat', 'open_yaml_files_cat']
```

**Comment by kthyng** [comment]
For now I am working around the 'open_erddap_cat' problem just calling it directly, though I'd appreciate knowing what to do about that issue.

I am hitting a different problem: I know I am making my catalog ok because if I stop my code in the middle of execution I can look at it and see entries and they look ok. But if I let it run to the end, then do something like `list(cat)`, I get:

```
File ~/packages/intake/intake/readers/mixins.py:41, in PipelineMixin.__getitem__(self, item)
     [36](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:36) outtype = self.output_instance
     [37](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:37) if "Catalog" in outtype:
     [38](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:38)     # a better way to mark this condition, perhaps the datatype's structure?
     [39](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:39)     # TODO: this prevents from doing a transform/convert on a cat, so must use
     [40](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:40)     #  .transform for that
---> [41](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:41)     return self.read()[item]
     [42](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:42) if isinstance(self, Pipeline):
     [43](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/mixins.py:43)     return self.with_step((GetItem, (item,), {}), out_instance=outtype)

File ~/packages/intake/intake/readers/entry.py:521, in Catalog.__getitem__(self, item)
    [519](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/entry.py:519)     return item.to_data(user_parameters=ups, **(kw or {}))
    [520](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/entry.py:520) else:
--> [521](https://file+.vscode-resource.vscode-cdn.net/Users/kthyng/projects/cook-inlet-catalogs/docs/demo_notebooks/~/packages/intake/intake/readers/entry.py:521)     raise KeyError(item)

KeyError: 0
```

I have dug into the code and can't figure out what to do about this. It looks like it is trying to call the catalog entries by number instead of by key, hence why it can't find one called 0, but I'm not sure. Any ideas?

**Comment by kthyng** [comment]
It seems like either `intake/intake/readers/mixins.py` or `intake/intake/readers/entry.py` needs to select `list(aliases)[item]` instead of just `item` to avoid looking for `0` instead of the keys themselves.

**Comment by martindurant** [comment]
intake.readers.Catalog has its own `__getitem__`, so you shouldn't be getting there. Are you sure `cat` is a catalog?

**Comment by kthyng** [comment]
I have `class ERDDAPCatalogReader(BaseReader)` and it has `output_instance = "intake.readers.entry:Catalog"`. I don't know why it does into `mixins.py` in the first place when I called `list(cat)` but once it does it checks `if "Catalog" in outtype` and the outtype is 'intake.readers.entry:Catalog'. Is there another way I should be setting it as a catalog?

**Comment by martindurant** [comment]
OK, so when you have an instance of ERDDAPCatalogReader, `.read()` would return the catalog, and you can do `list()` on that. It is possible to access into a catalog to get items with `.attribute` or [...] lookup, but there is no `iter` or whatever list() would need.

**Comment by kthyng** [comment]
Ahhh. Ok. I understand now. That makes sense once the pieces are in place but it's confusing going from v1 to v2 and trying to figure it out from my particular package.

I think the big pieces are almost in place other than how I am meant to call a v2 source and a v2 catalog that I have entry points for. For example, for the example we've been going through, set up this way:

```
    entry_points={
        "intake.imports": [
            "tabledap = intake_erddap.erddap:TableDAPSource",
            "griddap = intake_erddap.erddap:GridDAPSource",
            "erddap_cat = intake_erddap.erddap_cat:ERDDAPCatalogReader",
        ],
    },
```

what are the various ways I can refer to the Sources and Reader after importing intake without importing `intake-erddap` directly? (I guess I need to update the Sources to be called Data types or similar now.)

**Comment by martindurant** [comment]
If you look inside any catalog where you have included one of your data/readers (e.g., `cat.pprint()`), you will see that includes the package/module/class of the object directly. So, you don't actually need to use entrypoints or anything else at all. Ideally, the cat would include an environment in its metadata, so users know what must be installed to be able to make use of the cat.

Reusing the example from before:
```
In [11]: import intake_project
In [12]: cat["data"] = intake_project.ExampleData()
In [13]: cat.pprint()
{'version': 2,
 'data': {'data': {'datatype': 'intake_project:ExampleData',  # <- here
                   'kwargs': {},
                   'metadata': {},
                   'user_parameters': {}}},
 'aliases': {},
 'metadata': {},
 'user_parameters': {},
 'entries': {}}
```

We do not have the following features from Intake 1, which may we worth reimplementing:
- open_* functions based on all available things (but we do have recommendation functions which sort of do some of that)
- intake.cat , an obvious way to populate a global catalog with V2 objects (the V1 version of this still works)

Note so self: the config allows preventing auto-importing objects due to safety concerns. It would be reasonable to skip that check, should the target already be in memory.

Note to self: to reimplement `list(cat)` to show anything that is not an auto-generated hash

**Comment by kthyng** [comment]
Ok, thanks, that is useful.

What are the entry_points meant to be used for? There must be some connective tissue set up by them so that I can call something like:

`intake.[something].ERDDAPCatalogReader` and `intake.[something].TableDAPSource`

right? Otherwise what is the point of the entry_points and otherwise I would always just go directly through my `intake-erddap` package to access the Reader and Sources I would think.

**Comment by martindurant** [comment]
Adding the entrypoint means that the class is imported and will appear in the class hierarchy of the basic types. That means that .to_reader(), recommend, auto_pipeline, and other guessing functions know about it.

Looking up the "something" by name, given that it has been auto-imported is an interesting idea.
